### DISCUSSION TITLE
Provide Map alternatives for where currently Dictionary is used

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#545] [Ideas]
For example there is

- `org.osgi.framework.BundleContext.registerService(Class<S>, S, Dictionary<String, ?>)`
- `org.osgi.framework.ServiceRegistration.setProperties(Dictionary<String, ?>)`
- ...

I think it would be good to fade out Dictonary from the OSGi API and always provide alternatives that are using a map, for example one can have the follwoing default implemented method in the BundleContext interface:

```
default <S> ServiceRegistration<S> registerService(Class<S> clazz, S service, Map<String, ?> properties) {
     return registerService(clazz, service, FrameworkUtil.asDictionary(properties));
}
```

the old methods can be marked as deprecated and for removal so they can be removed some time in the future with a major version bump.

**Comment by bosschaert** [comment]
This issue has been brought up a number of times in the past at OSGi. At least up until now the viewpoint has always been that OSGi would never do a breaking change of its APIs and therefore that a major version increase will not happen, ergo the _deletion_ of the Dictionary from the OSGi API won't happen.

Supporting map overloads alongside the antiquated Dictionary ones should be possible, but the question is whether its worth the effort as it doesn't really add that much...

**Reply by tjwatson** [reply]
Adding overloaded methods also causes churn for code that uses implementations that extend Dictionary and implement Map (e.g. Hashtable) because now the method becomes ambiguous.  While not considered a binary breaking change, it can be considered a source code breaking change.  We have past precedent where we broke source code compatibility, forcing code to change if it wants to compile against the latest OSGi APIs. But I do think this type of overload could cause large amounts of existing code to change in order to compile against the latest OSGi APIs.

Not a blocker to making the change, but I question the overall value of doing this for system level APIs that we want 90%+ of developers to avoid ever using.

**Reply by juergen-albert** [reply]
I concur that it does not add some real technical value. I think however that providing some new paint to our APIs that use more modern and nowadays commonly used signatures will not hurt to make us a bit more attractive. Your example however sounds strange but valid and I see the issue here.

| Not a blocker to making the change, but I question the overall value of doing this for system level APIs that we want 90%+ of developers to avoid ever using.

We want to encourage people to use Components and not register Services in Activators. But when you really work with services registering some Things manually is not such an uncommon case I think. It is actually unavoidable, if I want a service, that can control its own properties to mark e.g. states.

**Comment by bjhargrave** [comment]
This idea is always raised. Dictionary is not deprecated, it is just not recommended. For most OSGi users that use DS (as we recommend), the use of Map is already well supported. Only users working with the low level framework API will need to interact with Dictionary. And we now have FrameworkUtil.asDictionary if they need to use a pre-existing Map. But normally, the user will new a Hashtable which is both a Map and Dictionary. And, as Tom observed, overloading methods which take Dictionary with Map variants will lead to compilation failures.

So I am -1 to do this since it does not add real value, it bloats the API, and it risks source compatibility issues.

**Reply by laeubi** [reply]
Dictionary is not deprecated but a specialized (abstract) type that only has one implementation in java that code scanners mark as "often not required" and it hardly interoperates with any slightly modern java API.
Beside that, the compile error is easy to fix (e.g. by just declaring the type as Dictonary), so this more shows how Dictionary was a bad choice in the past, but this does not mean one can't change in the future.

All this "we want to encourage X" is actually irrelevant I think, as I suppose the choice for `Dictonary` was not meant to annoy people to not use the OSGi-API, and I have never heard about anyone changing to DI because of a `Dictonary` parameter ...

Beside that there are still use-cases where this is required and thus I think it is valuable to change, especially if the topic was brought up multiple times... `Map` is there since **Java 1.2 (!)** more than 25 years ago now so things have really changed in between those days... This just makes people feel the OSGi API is old and outdated and no one really cares about it...

**Reply by mikerumpf** [reply]
I concur with Christoph. This should be updated.  While we use ds as much as we can there are still some cases where we need to use the API programmatically. I have been asked many times why on earth OSGi still uses Dictionary. Doesn't really feel like a modern API.

**Reply by stbischof** [reply]
@mikerumpf could you please explain in which cases you have to handle the Dictionary?

**Reply by mikerumpf** [reply]
Yes of course: There are several use cases: We are reading OSGi-configurations from non-file-locations and have to access the ConfigurationAdmin for this purpose. Also our client is an E4/RCP-application. There are places where we need to register services someplace in the E4 lifecycle because there is no other way to to that. Also within ds we use the ComponentFactory API which also still needs a Dictionary.

**Comment by laeubi** [comment]
If overloading won't work maybe one can support either:

- an alternative name (e.g. `org.osgi.framework.BundleContext.registerServiceWithProperties(Class<S>, S, Map<String, ?>)`)
- signatures similar to what Map.of(...) offers e.g. `org.osgi.framework.BundleContext.registerServiceWithProperties(Class<S>, S, String key, Object val)`

in every seldom case I need to call such API its always feels like brought back to Java 1.4 days...

**Reply by stbischof** [reply]
This is not the only place. Also configurator.

Is there any Implementation that hat such a method and will Support it?
